// material-ui-popup-state/HoverMenu.js
"use strict";

var _interopRequireDefault2 = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault2(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault2(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var _material = require("@mui/material");
var _interopRequireDefault = require("./util/interopRequireDefault.js");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, "default": e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /* eslint-disable @typescript-eslint/no-deprecated */
// work around https://github.com/vercel/next.js/issues/57285
var Menu = (0, _interopRequireDefault.interopRequireDefault)(_material.Menu);
var HoverMenu = /*#__PURE__*/React.forwardRef(function HoverMenu(props, ref) {
  var _props$slotProps2, _props$PaperProps;
  var paperSlotProps = React.useMemo(function () {
    var _props$slotProps;
    var wrapped = (_props$slotProps = props.slotProps) === null || _props$slotProps === void 0 ? void 0 : _props$slotProps.paper;
    if (wrapped instanceof Function) {
      return function (ownerProps) {
        var base = wrapped(ownerProps);
        return _objectSpread(_objectSpread({}, base), {}, {
          style: _objectSpread({
            pointerEvents: 'auto'
          }, base.style)
        });
      };
    }
    return _objectSpread(_objectSpread({}, wrapped), {}, {
      style: _objectSpread({
        pointerEvents: 'auto'
      }, wrapped === null || wrapped === void 0 ? void 0 : wrapped.style)
    });
  }, [(_props$slotProps2 = props.slotProps) === null || _props$slotProps2 === void 0 ? void 0 : _props$slotProps2.paper]);
  return /*#__PURE__*/React.createElement(Menu, (0, _extends2["default"])({}, props, {
    ref: ref,
    style: _objectSpread({
      pointerEvents: 'none'
    }, props.style),
    PaperProps: _objectSpread(_objectSpread({}, props.PaperProps), {}, {
      style: _objectSpread({
        pointerEvents: 'auto'
      }, (_props$PaperProps = props.PaperProps) === null || _props$PaperProps === void 0 ? void 0 : _props$PaperProps.style)
    }),
    slotProps: _objectSpread(_objectSpread({}, props.slotProps), {}, {
      paper: paperSlotProps
    })
  }));
});
var _default = exports["default"] = HoverMenu;
module.exports = exports.default;
